File: backend/services/storageService.js
import { PutObjectCommand, DeleteObjectCommand } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import { s3Client } from "../config/r2.js";
import crypto from "crypto";
import Logger from "../utils/logger.js";

/**
 * Service untuk mengelola interaksi dengan Cloudflare R2
 */
export const storageService = {
  /**
   * Men-generate Presigned URL agar Frontend bisa upload langsung ke R2
   * @param {string} originalName - Nama file asli
   * @param {string} mimeType - Content-Type file (e.g. image/jpeg)
   * @param {string} folder - Folder tujuan (e.g. 'profiles', 'products')
   * @returns {Promise<{url: string, key: string, publicUrl: string}|null>}
   */
  generatePresignedUploadUrl: async (originalName, mimeType, folder = "uploads") => {
    if (!s3Client) {
      throw new Error("S3 Client belum diinisialisasi. Periksa kredensial R2.");
    }

    try {
      // Generate nama file unik agar tidak bentrok
      const ext = originalName.split(".").pop();
      const randomString = crypto.randomUUID();
      const timestamp = Date.now();
      const uniqueFileName = `${folder}/${timestamp}-${randomString}.${ext}`;

      const command = new PutObjectCommand({
        Bucket: process.env.R2_BUCKET_NAME,
        Key: uniqueFileName,
        ContentType: mimeType,
      });

      // URL valid selama 5 menit
      const signedUrl = await getSignedUrl(s3Client, command, { expiresIn: 300 });

      // Hitung URL publik berdasarkan custom domain
      const publicUrl = `${process.env.R2_PUBLIC_URL}/${uniqueFileName}`;

      return {
        url: signedUrl,
        key: uniqueFileName,
        publicUrl: publicUrl,
      };
    } catch (error) {
      Logger.error("Gagal membuat presigned URL", error, "STORAGE_SERVICE");
      return null;
    }
  },

  /**
   * Menghapus object dari Cloudflare R2
   * @param {string} key - R2 Object Key (e.g. 'main/filename.webp')
   * @returns {Promise<boolean>}
   */
  deleteFromR2: async (key) => {
    if (!s3Client) {
      Logger.warn("S3 Client belum diinisialisasi. Lewati penghapusan dari R2.", "STORAGE_SERVICE");
      return false;
    }
    if (!key) return false;

    try {
      const command = new DeleteObjectCommand({
        Bucket: process.env.R2_BUCKET_NAME,
        Key: key,
      });
      await s3Client.send(command);
      return true;
    } catch (error) {
      Logger.error(`Gagal menghapus file dari R2: ${key}`, error, "STORAGE_SERVICE");
      return false;
    }
  }
};

File: backend/internal/service/storage_service.go
package service

import (
	"context"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/dps-wmhris/backend/internal/config"
	"github.com/google/uuid"
)

type StorageService interface {
	GeneratePresignedUploadUrl(ctx context.Context, originalName, mimeType, folder string) (url, key, publicUrl string, err error)
	DeleteFromR2(ctx context.Context, key string) (bool, error)
	UploadFile(ctx context.Context, fileContent []byte, originalName, mimeType, folder string) (publicUrl string, err error)
}

type storageServiceImpl struct{}

func NewStorageService() StorageService {
	return &storageServiceImpl{}
}

func (s *storageServiceImpl) GeneratePresignedUploadUrl(ctx context.Context, originalName, mimeType, folder string) (string, string, string, error) {
	if config.R2PresignClient == nil {
		return "", "", "", fmt.Errorf("S3 Client belum diinisialisasi. Periksa kredensial R2")
	}

	if folder == "" {
		folder = "uploads"
	}

	parts := strings.Split(originalName, ".")
	ext := parts[len(parts)-1]
	if len(parts) == 1 {
		ext = "" // no extension
	}

	randomStr := uuid.New().String()
	timestamp := time.Now().UnixMilli()

	uniqueFileName := fmt.Sprintf("%s/%d-%s", folder, timestamp, randomStr)
	if ext != "" {
		uniqueFileName += "." + ext
	}

	bucketName := os.Getenv("R2_BUCKET_NAME")

	req, err := config.R2PresignClient.PresignPutObject(ctx, &s3.PutObjectInput{
		Bucket:      aws.String(bucketName),
		Key:         aws.String(uniqueFileName),
		ContentType: aws.String(mimeType),
	}, func(po *s3.PresignOptions) {
		po.Expires = 5 * time.Minute
	})

	if err != nil {
		log.Printf("[STORAGE_SERVICE] Gagal membuat presigned URL: %v", err)
		return "", "", "", err
	}

	publicUrl := fmt.Sprintf("%s/%s", os.Getenv("R2_PUBLIC_URL"), uniqueFileName)

	return req.URL, uniqueFileName, publicUrl, nil
}

func (s *storageServiceImpl) DeleteFromR2(ctx context.Context, key string) (bool, error) {
	if config.R2Client == nil {
		log.Println("[STORAGE_SERVICE] S3 Client belum diinisialisasi. Lewati penghapusan dari R2")
		return false, nil
	}
	if key == "" {
		return false, nil
	}

	bucketName := os.Getenv("R2_BUCKET_NAME")
	_, err := config.R2Client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(bucketName),
		Key:    aws.String(key),
	})

	if err != nil {
		log.Printf("[STORAGE_SERVICE] Gagal menghapus file dari R2 %s: %v", key, err)
		return false, err
	}
	return true, nil
}

func (s *storageServiceImpl) UploadFile(ctx context.Context, fileContent []byte, originalName, mimeType, folder string) (string, error) {
	if config.R2Client == nil {
		return "", fmt.Errorf("S3 Client belum diinisialisasi")
	}

	if folder == "" {
		folder = "exports"
	}

	parts := strings.Split(originalName, ".")
	ext := parts[len(parts)-1]
	if len(parts) == 1 {
		ext = ""
	}

	randomStr := uuid.New().String()
	timestamp := time.Now().UnixMilli()

	uniqueFileName := fmt.Sprintf("%s/%d-%s", folder, timestamp, randomStr)
	if ext != "" {
		uniqueFileName += "." + ext
	}

	bucketName := os.Getenv("R2_BUCKET_NAME")

	_, err := config.R2Client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:      aws.String(bucketName),
		Key:         aws.String(uniqueFileName),
		Body:        strings.NewReader(string(fileContent)),
		ContentType: aws.String(mimeType),
	})

	if err != nil {
		log.Printf("[STORAGE_SERVICE] Gagal mengupload file ke R2: %v", err)
		return "", err
	}

	publicUrl := fmt.Sprintf("%s/%s", os.Getenv("R2_PUBLIC_URL"), uniqueFileName)
	return publicUrl, nil
}
